doc/designs/black-codestyle.rst

```diff
@@ -0,0 +1,308 @@
+
+then it's acceptable to disable auto-formatting.
+
+.. code-block:: python
+
+   # fmt: off
+   function(
+       argument_a, argument_b, argument_c, argument_d, argument_e,
+       argument_f, argument_g, argument_h,
+   )
+   # fmt: on
+
+.. _black-pycodestyle:
+
+pycodestyle
+===========
+
+With |Black| auto-formatting and a handful of minor patches it is finally
+possible to run `pycodestyle`_ successfully on the entire code base.
+Remaining style issues were already addressed in upstream `PR 4638`_.
+
+Fixed Style issues
+------------------
+
+* *E266* too many leading '#' for block comment
+* *E302* expected 2 blank lines, found 1
+* *W601* .has_key() is deprecated, use 'in'
+* *E711* comparison to None should be 'if cond is None:'
+* *E712* comparison to True should be 'if cond is True:' or 'if cond:'
+* *E712* comparison to False should be 'if cond is False:' or 'if not cond:'
+* *E713* test for membership should be 'not in'
+* *E714* test for object identity should be 'is not'
+* *E721* do not compare types, use 'isinstance()'
+* *E722* do not use bare 'except'
+
+New ignores
+-----------
+
+*E203* whitespace before ':'
+   *E203* is not :pep:`8` conform. |Black| treats slice ``:`` as binary
+   operator and enforces whitespace in slices, for example ``ham[1 + 1 :]``.
+*E231* missing whitespace after ','
+   |Black| always adds a comma after all arguments, e.g. ``func(a,)``.
+*W503* line break before binary operator
+   *W503* is not :pep:`8` conform.
+*W504* line break after binary operator
+   In rare cases |Black| adds a binary operator on its own line when an
+   expression contains inline comments.
+*E731* do not assign a lambda expression
+   IPA creates callable from lambdas a lot. It doesn't make sense to change
+   all places.
+*E741* ambiguous variable name 'l'
+   In several places IPA uses ``l`` as variable name. In some fonts it can
+   be confused with number ``1``.
+
+Implementation
+==============
+
+1. Create infrastructure for |Black|
+
+   * Add ``BuildRequires: black`` (|Black| is available in Fedora)
+   * Add ``make`` targets ``black`` and ``blacklint``
+   * Create ``pyproject.toml`` to configure |Black| and include Python code
+     that does not have a ``.py`` file extension.
+   * Exclude auto-generated plugin code in ``ipaclient/remote_plugins/2_???``
+     from black. It's legacy code and no developer is going to touch the code
+     any more.
+
+2. Address remaining :ref:`pycodestyle issues <black-pycodestyle>` by either
+   fixing the issue or ignoring the warning locally or globally.
+   ``python3 -m pycodestyle .`` should pass without any error.
+3. Backport changes from (1) and (2) to ipa-4-8 branch.
+4. Run ``make black`` in master + ipa-4-8 branch and merge the changes.
+5. Create ``.gitignorerevs`` file with commit hashes of |Black| run from all
+   active branches``.
+6. Enable ``blacklint`` for ``fastlint`` and ``lint`` targets so
+   local linting and linting on Azure check for black violations.
```

Comment:
rather than running blacklint in CI, if we know all the code is black-compliant, then add a git commit check to require black formatting so a PR can't be submitted at all w/o complying with formatting. This way users wouldn't have to wait for CI to find out they have a formatting issue. I don't always run fastlint myself before submitting a PR.